src/java.base/share/classes/java/util/zip/DeflaterOutputStream.java

```diff
@@ -240,14 +240,26 @@ public void finish() throws IOException {
      */
     public void close() throws IOException {
         if (!closed) {
+            IOException finishException = null;
             try {
                 finish();
+            } catch (IOException ioe){
+                finishException = ioe;
+                throw ioe;
             } finally {
-                if (usesDefaultDeflater)
+                if (usesDefaultDeflater) {
                     def.end();
+                }
+                try {
+                    out.close();
+                } catch (IOException ioe) {
+                    if (finishException != ioe) {
+                        ioe.addSuppressed(finishException);
```

Comment:
A null check for `finishException` will be needed here to prevent a `NullPointerException` being thrown from within `addSuppressed`. I think it's better to just copy over (rest of) the finally block from `FilterOutputStream`'s close() method.

Reply:
Updated the code to align with `FilterOutputStream.close()`:

- Avoid the suppression try/catch for the case where finishException is null
- Move `closed = true;` to the top of the method, this makes sure the wrapped stream is only closed once even in the case where it throws during close. Close should only be attempted once.

I added more test cases including one where both finish and close operations fail, one where finish and close operations throw identical IOException instances and one testing that the wrapped stream is closed only once, even if the close operation throws.